[AA] Refine ModRefInfo taking into account `errnomem` location

antoniofrighetto opened the pull request:
Ensure alias analyses mask out `errnomem` location, refining the resulting modref info, when the given access/location does not alias errno. This may occur either when TBAA proves there is no alias with errno (e.g., float TBAA for the same root would be disjoint with the int-only compatible TBAA node for errno); or if the memory access size is larger than the integer size, or when the underlying object is a potentially-escaping alloca.

Related PR: https://github.com/llvm/llvm-project/pull/125258.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

dtcxzyw commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -997,6 +998,18 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+  // the given memory location is an alloca, the size of the memory access is
+  // larger than the integer size, or if TBAA proves it does not alias errno.
+  if ((ErrnoMR | Result) != Result) {
+    if (ErrnoMR == ModRefInfo::Mod && !isa<AllocaInst>(Object) &&
```
```suggestion
    if (ErrnoMR == ModRefInfo::Mod && !isIdentifiedFunctionLocal(Object) &&
```

nikic replied:
Hm, I think identified function-local is not completely right here because it also includes noalias arguments. All of alloca, noalias call and byval are fine, but I believe a noalias argument can alias errno.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -951,7 +951,8 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
   ModRefInfo ArgMR = ME.getModRef(IRMemLocation::ArgMem);
-  ModRefInfo OtherMR = ME.getWithoutLoc(IRMemLocation::ArgMem).getModRef();
+  ModRefInfo ErrnoMR = ME.getModRef(IRMemLocation::ErrnoMem);
+  ModRefInfo OtherMR = ME.getModRef(IRMemLocation::Other);
```
Because a MemoryLocation cannot alias inaccessible memory. (Note that it is already explicitly excluded a few lines above.)

antoniofrighetto commented on llvm/test/Transforms/EarlyCSE/writeonly.ll:
```diff
@@ -17,7 +17,7 @@ define void @test() {
 }
 
-declare void @writeonly_void() memory(write)
+declare void @writeonly_void() memory(write, errnomem: none)
```
We were treating noalias arguments as may alias errno thus writeonly_cse_intervening_noalias_store was failing, now reverted this.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -997,6 +998,19 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+  // of the memory access is larger than the integer size, or if TBAA proves it
+  // does not alias errno.
+  if ((ErrnoMR | Result) != Result) {
+    if (ErrnoMR == ModRefInfo::Mod && !isIdentifiedFunctionLocal(Object) &&
```
Why is this limited to `ErrnoMR == ModRefInfo::Mod`? Shouldn't this logic apply to both mod and ref of errno?

antoniofrighetto commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -997,6 +998,19 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+  // of the memory access is larger than the integer size, or if TBAA proves it
+  // does not alias errno.
+  if ((ErrnoMR | Result) != Result) {
+    if (isModSet(ErrnoMR) && !isIdentifiedFunctionLocal(Object) &&
```
Hmm, right. Somehow I thought we could ignore reads (due to how we conceived errno), except that ignoring them would just lead to a less sound result. Thanks, updated, added another test as well.

nikic commented on llvm/test/Transforms/InstCombine/may-alias-errno.ll:
```diff
@@ -0,0 +1,133 @@
+}
+
+declare float @sinf(float) memory(errnomem: write)
+declare float @cosf(float) memory(argmem: write, errnomem: read)
```
I'd use a function other than cosf for this. It's confusing, as the actual cosf wouldn't have these memory effects...

antoniofrighetto replied:
Done, thanks.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -997,6 +998,18 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+  // Refine accesses to errno memory. We can safely exclude the errno location
+  // if the given memory location is an identified function-local object, the
+  // size of the memory access is larger than the integer size, or if TBAA
+  // proves it does not alias errno.
```
I think most of this comment belongs into aliasErrno() now.

nikic reviewed (approved):
LGTM

antoniofrighetto commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -997,6 +998,19 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+  // size of the memory access is larger than the integer size, or if TBAA
+  // proves it does not alias errno.
+  if ((ErrnoMR | Result) != Result) {
+    if (isModOrRefSet(ErrnoMR) && !isIdentifiedFunctionLocal(Object) &&
```
Right. Kept it there for compile-time concerns, moved it in aliasErrno.

antoniofrighetto committed:
[AA] Refine ModRefInfo taking into account `errnomem` location

Ensure alias analyses mask out `errnomem` location, refining the
resulting modref info, when the given access/location does not
alias errno. This may occur either when TBAA proves there is no
alias with errno (e.g., float TBAA for the same root would be
disjoint with the int-only compatible TBAA node for errno); or
if the memory access size is larger than the integer size, or
when the underlying object is a potentially-escaping alloca.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

nikic commented:
@madhur13490 Are you still seeing this after https://github.com/llvm/llvm-project/commit/09966960c692be63b192448e8c2b6bfebad3c3b8? That should have fixed it.
